### PR TITLE
Implement batch API with changeset, upsert, and DataFrame integration

### PR DESCRIPTION
**Pull request by sagebree**
## Summary

- Adds `client.batch` namespace -- a deferred-execution batch API that packs multiple
  Dataverse Web API operations into a single `POST $batch` HTTP request
- Adds `client.batch.dataframe` namespace -- pandas DataFrame wrappers for batch operations
- Adds `client.records.upsert()` and `client.batch.records.upsert()` backed by the
  `UpsertMultiple` bound action with alternate-key support
- Fixes a bug where alternate key fields were merged into the UpsertMultiple request
  body, causing `400 Bad Request` on the create path

## Batch API Design

Implements the [Batch API Design](https://github.com/microsoft/PowerPlatform-DataverseClient-Python/pull/129#issuecomment-3974570865) spec from @sagebree:

| Capability | How to use | Status |
|---|---|---|
| Record CRUD (create / update / delete / get) | `batch.records.*` | Done |
| Upsert by alternate key | `batch.records.upsert(...)` | Done |
| Table metadata (create / delete / columns / relationships) | `batch.tables.*` | Done |
| SQL queries | `batch.query.sql(...)` | Done |
| Atomic write groups | `batch.changeset()` | Done |
| Continue past failures | `batch.execute(continue_on_error=True)` | Done |
| DataFrame integration | `batch.dataframe.create/update/delete` | Done (new) |

**Design constraints enforced:**
- Maximum 1000 operations per batch (validated before sending)
- `records.get` paginated overload not supported -- single-record only
- GET operations cannot be placed inside a changeset (enforced by API design)
- Content-ID references are only valid within the same changeset
- File upload operations not batchable
- `tables.create` returns no table metadata on success (HTTP 204)
- `tables.add_columns` / `tables.remove_columns` do not flush the picklist cache
- `client.flush_cache()` not supported in batch (client-side operation)

### New: `client.batch` API
- `batch.records.create / get / update / delete / upsert`
- `batch.tables.create / get / list / add_columns / remove_columns / delete`
- `batch.tables.list(filter=..., select=...)` -- parity with `client.tables.list()` from #112
- `batch.tables.create_one_to_many_relationship / create_many_to_many_relationship / delete_relationship / get_relationship / create_lookup_field`
- `batch.query.sql`
- `batch.changeset()` context manager for transactional (all-or-nothing) operations
- Content-ID reference chaining inside changesets (globally unique across all changesets via shared counter)
- `execute(continue_on_error=True)` for mixed success/failure batches
- `BatchResult` with `.responses`, `.succeeded`, `.failed`, `.created_ids`, `.has_errors`

### New: `client.batch.dataframe` API
- `batch.dataframe.create(table, df)` -- DataFrame rows to CreateMultiple batch item
- `batch.dataframe.update(table, df, id_column)` -- DataFrame rows to update batch items
- `batch.dataframe.delete(table, ids_series)` -- pandas Series to delete batch items

### Existing: Refactored existing APIs
- Payload generation shared between batch and direct API via `_build_*` / `_RawRequest` pattern
- Execution of batch operations deferred to `execute()`

### OData $batch spec compliance
- Audited against [Microsoft Learn docs](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/execute-batch-operations-using-web-api)
- `Content-Transfer-Encoding: binary` per part
- `Content-Type: application/http` per part
- `Content-Type: application/json; type=entry` for POST/PATCH bodies
- CRLF line endings throughout
- Absolute URLs in batch parts
- Empty changesets silently skipped (prevents invalid multipart)
- Top-level batch error handling (non-multipart 4xx/5xx raises `HttpError` with parsed Dataverse error details)
- Accepts `200`, `202 Accepted`, `207 Multi-Status`, and `400` batch response codes

### Review comment fixes
- Fixed `expected` status codes to include `202`/`207` for all Dataverse environments
- Fixed `_split_multipart` / `_parse_mime_part` return type annotations: `List[Tuple[Dict[str, str], str]]`
- Fixed OptionSet string check regression: now uses dict key lookup instead of JSON string search
- Fixed `_build_get` to lowercase select column names (consistency with `_get_multiple`)
- Added RFC 3986 `%20` encoding documentation in `_build_sql` docstring
- Fixed content-id response parsing for non-changeset parts
- Fixed test assertions after merge: `data` bytes instead of `json` kwarg
- Exception type parity: `batch.records.upsert()` raises `TypeError` (matching `client.records.upsert()`)

### Testing

**Unit tests -- 579 tests passing:**
- `test_batch_operations.py` -- BatchRequest, BatchRecordOperations, BatchTableOperations, BatchQueryOperations, ChangeSet, BatchItemResponse, BatchResult
- `test_batch_serialization.py` -- multipart serialization, response parsing, intent resolution, upsert dispatch, batch size limit, content-ID uniqueness, top-level error handling
- `test_batch_edge_cases.py` -- 40 edge case tests: empty changeset, changeset rollback, content-ID in standalone parts, mixed batch, multiple changesets, batch size limits, top-level errors, continue-on-error, serialization compliance, multipart parsing, content-ID references, intent validation
- `test_batch_dataframe.py` -- 18 tests: DataFrame create/update/delete, validation, NaN handling, empty series, bulk delete
- `test_odata_internal.py` -- `_build_upsert_multiple` body exclusion, conflict detection, URL/method correctness

**E2E tests -- 14 tests passing against live Dataverse (`crm10.dynamics.com`):**
1. Basic batch CRUD (single create + CreateMultiple, update, get, delete)
2. Changeset happy path (create + update via `$ref` content-ID)
3. Changeset rollback (failing op rolls back entire changeset)
4. Multiple changesets (globally unique content-IDs)
5. Continue-on-error (mixed success/failure)
6. Batch SQL query
7. Batch tables.get + tables.list
8. DataFrame batch create
9. DataFrame batch update
10. DataFrame batch delete
11. Mixed batch (changeset + standalone GET)
12. Empty changeset (silently skipped)
13. Content-ID chaining (2 creates + 2 updates via `$ref`)
14. Table setup/teardown

### Examples & docs
- `examples/advanced/batch.py` -- reference examples for all batch operation types
- `examples/advanced/walkthrough.py` -- batch section added (section 11)
- `examples/basic/functional_testing.py` -- `test_batch_all_operations()` covering all operation categories against a live environment

### CODE REVIEW

File: src/PowerPlatform/Dataverse/data/_odata.py [hunk 1]
```diff
@@ -1685,6 +1586,412 @@ def _delete_columns(
...
+        ``?sql=`` query parameter.
+
+        :param sql: SELECT statement (non-empty string; caller is responsible
+            for validation).
+        """
+        logical = self._extract_logical_table(sql)
+        entity_set = self._entity_set_from_schema_name(logical)
+        return _RawRequest(
+            method="GET",
+            url=f"{self.api}/{entity_set}?sql={_url_quote(sql, safe='')}",
```
Comment (tpellissier-msft) [inline src/PowerPlatform/Dataverse/data/_odata.py]: Before, we could just do

url = f"{self.api}/{entity_set}"
params = {"sql": sql}
r = self._request("get", url, params=params)

And the requests library would handle the URL encoding. But now with using RawRequest, we have to do

 url=f"{self.api}/{entity_set}?sql={_url_quote(sql, safe='')}

and encode the SQL ourselves. This puts more security burden on us and I'm not sure if the encoding would be the same. Claude suggests it might not:

"This changes encoding behavior — requests uses urllib.parse.urlencode which encodes spaces as +, while _url_quote(safe='') encodes spaces as %20. Is this intentional? Has he tested SQL with special characters (spaces, =, &, /) against a live environment? The unit tests in test_sql_parse.py test the encoding shape but not against Dataverse."

File: src/PowerPlatform/Dataverse/data/_batch.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+# Changeset container
+# ---------------------------------------------------------------------------
+
+
+@dataclass
+class _ChangeSet:
+    """Ordered group of single-record write operations that execute atomically."""
+
+    operations: List[Union[_RecordCreate, _RecordUpdate, _RecordDelete]] = field(default_factory=list)
+    _next_content_id: int = field(default=1, init=False, repr=False)
```
Comment (tpellissier-msft) [inline src/PowerPlatform/Dataverse/data/_batch.py]: Claude raised this question about multiple changesets and _next_content_id that I didn't know the answer to, so passing it on to you :)

"Each _ChangeSet starts its content-ID counter at 1. If a batch has two changesets, both will have Content-ID: 1, Content-ID: 2, etc. The OData spec says content-IDs must be unique within a batch (not just within a changeset). Dataverse may or may not enforce this — worth confirming."

File: tests/unit/data/test_sql_parse.py [hunk 3]
```diff
@@ -1,6 +1,10 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT license.
 
+import unittest
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply tests/unit/data/test_sql_parse.py]: Verified -- there is no bare import unittest in this file. The only import is from unittest.mock import patch which is actively used. No change needed.

[the remark this answers is not in the document]
Comment (saurabhrb) [reply tests/unit/data/test_sql_parse.py]: Verified -- there is no bare import unittest in this file. The only import is from unittest.mock import patch which is actively used. No change needed.

File: src/PowerPlatform/Dataverse/data/_odata.py (hunk 1 above)
Comment (saurabhrb) [reply src/PowerPlatform/Dataverse/data/_odata.py]: Good catch. Added an explicit docstring note in _build_sql (537b198) documenting that urllib.parse.quote (%20 for spaces) is the intentional choice rather than urlencode (+ for spaces). Both are accepted by Dataverse, and %20 is the canonical RFC 3986 encoding for query-string values. The e2e tests in .scratch/e2e_batch_test.py (test 6) confirm SQL with special characters works correctly against a live Dataverse environment.

File: src/PowerPlatform/Dataverse/data/_batch.py [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+        if continue_on_error:
+            headers["Prefer"] = "odata.continue-on-error"
+
+        url = f"{self._od.api}/$batch"
+        response = self._od._request(
+            "post",
+            url,
+            data=body.encode("utf-8"),
+            headers=headers,
+            expected=(200, 400),
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply src/PowerPlatform/Dataverse/data/_batch.py]: Fixed in 537b198. Updated to expected=(200, 202, 207, 400) to cover all Dataverse batch response variants.

File: src/PowerPlatform/Dataverse/data/_odata.py [hunk 5]
```diff
@@ -1731,17 +1639,21 @@ def _create_columns(
...
-
-            created.append(column_name)
-
-            if "OptionSet" in payload:
+            attr = self._attribute_payload(column_name, column_type)
+            if not attr:
+                raise ValidationError(
+                    f"Unsupported column type '{column_type}' for column '{column_name}'.",
+                    subcode="unsupported_column_type",
+                )
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply src/PowerPlatform/Dataverse/data/_odata.py]: Fixed in 26c7437. All 3 occurrences now use VALIDATION_UNSUPPORTED_COLUMN_TYPE constant from core/_error_codes.py.

File: .claude/skills/dataverse-sdk-use/SKILL.md [hunk 6]
```diff
@@ -369,6 +370,50 @@ client.files.upload(
...
+result = batch.execute(continue_on_error=True)
+print(f"Succeeded: {len(result.succeeded)}, Failed: {len(result.failed)}")
+```
+
+**BatchResult properties:**
+- `result.responses` -- list of `BatchItemResponse` in submission order
+- `result.succeeded` -- responses with 2xx status codes
+- `result.failed` -- responses with non-2xx status codes
+- `result.has_errors` -- True if any response failed
+- `result.entity_ids` -- GUIDs from successful create operations
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply .claude/skills/dataverse-sdk-use/SKILL.md]: Fixed in f3c7b7e. Same fix applied.

File: src/PowerPlatform/Dataverse/operations/batch.py [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+            ref = cs.records.create("contact", {"firstname": "Alice"})
+            cs.records.update("account", account_id, {
+                "primarycontactid@odata.bind": ref
+            })
+        result = batch.execute()
+    """
+
+    def __init__(self, client: "DataverseClient") -> None:
+        self._client = client
+        self._items: list = []
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply src/PowerPlatform/Dataverse/operations/batch.py]: Fixed in f3c7b7e. Changed to List[Any].

File: src/PowerPlatform/Dataverse/operations/batch.py [hunk 8]
```diff
@@ -0,0 +1,829 @@
...
+    LookupAttributeMetadata,
+    OneToManyRelationshipMetadata,
+    ManyToManyRelationshipMetadata,
+)
+from ..common.constants import CASCADE_BEHAVIOR_REMOVE_LINK
+
+if TYPE_CHECKING:
+    from ..client import DataverseClient
+
+__all__ = []
```
Comment (abelmilash-msft) [inline src/PowerPlatform/Dataverse/operations/batch.py]: Nit: `__all__` is empty. It seems like there is a pattern of exporting using `__all__` in other modules. May be worth keeping that pattern here.

File: examples/advanced/batch.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+# Licensed under the MIT license.
+
+"""
+Batch operations example for the Dataverse Python SDK.
+
+Demonstrates how to use client.batch to send multiple operations in a single
+HTTP request to the Dataverse Web API.
+
+Requirements:
+    pip install "PowerPlatform.Dataverse" azure-identity
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply examples/advanced/batch.py]: Fixed in 33a45bc. Updated to `pip install PowerPlatform-Dataverse-Client azure-identity`.

File: tests/unit/data/test_batch_serialization.py [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+
+        od._build_get.assert_called_once_with("account", "guid-1", select=["name"])
+        self.assertEqual(result, [mock_req])
+
+    def test_resolve_record_delete_single(self):
+        client, od = self._client_and_od()
+        mock_req = MagicMock()
+        od._build_delete.return_value = mock_req
+
+        op = _RecordDelete(table="account", ids="guid-1")
```
[the remark this answers is not in the document]
Comment (saurabhrb) [reply tests/unit/data/test_batch_serialization.py]: Fixed in 33a45bc. Added tests for: multiple IDs with bulk delete, multiple IDs without bulk delete (individual requests), empty list returns empty, filtering out empty strings from ID list, and all-empty-strings returns empty.
